model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -379,38 +306,136 @@ func (mm *OvmsModelManager) run() {
+	for {
+		select {
+		case <-stopChan:
+			return requestMap
+
+		case req, ok := <-mm.requests:
+			if !ok {
+				// shutdown the run() loop after processing this batch
+				mm.requests = nil
+				break // from select
+			}
+
+			// has the context been cancelled?
+			if err := req.ctx.Err(); err != nil {
+				s := status.FromContextError(err)
+				mm.log.V(1).Info("Aborting request with cancelled context", "model_id", req.modelId, "request_type", req.requestType, "code", s.Code())
+				completeRequest(req, s.Code(), "Request context has been cancelled")
+				continue
+			}
+
 			switch req.requestType {
-			case load:
-				if !statusExists {
-					completeRequest(req, codes.Internal, "Expected model to load, but no status entry found")
-				} else if modelStatus.State == "AVAILABLE" {
-					completeRequest(req, codes.OK, "")
-				} else {
-					completeRequest(req, codes.InvalidArgument, fmt.Sprintf("OVMS model load failed. code: '%s' reason: '%s'", modelStatus.Status.ErrorCode, modelStatus.Status.ErrorMessage))
+			case unloadAll:
+				mm.log.V(1).Info("Processing UnloadAll", "numRequests", len(requestMap))
+
+				// abort all pending requests
+				for _, r := range requestMap {
+					mm.log.V(1).Info("Aborting request to reset the Model Server", "model_id", r.modelId, "request_type", r.requestType, "context_error", r.ctx.Err().Error())
+					completeRequest(r, codes.Aborted, "Model Server is being reset")
 				}
+				requestMap = map[string]*request{}
+
+				// an UnloadAll does not need to trigger a reload right now, we can report success
+				// and will sync state with the model server on the next reload
+				completeRequest(req, codes.OK, "")
+
+				// reset the desired model state to be empty
+				mm.loadedModelsMap = map[string]OvmsMultiModelConfigListEntry{}
+
+				// reset the stop timer
+				if stopChan == nil {
+					shortTimerSet = false
+					stopChan = time.NewTimer(mm.config.BatchWaitTimeMax).C
+				}
+
 			case unload:
-				if !statusExists {
-					// OVMS keeps an entry for a model that has been unloaded, so this case means that the
-					// model was not loaded previously. This is unexpected, but we can proceed
-					log.Info("Processed UnloadModel for model that was never loaded", "modelId", req.modelId)
-					completeRequest(req, codes.OK, "")
-				} else if modelStatus.State == "END" {
-					completeRequest(req, codes.OK, "")
-				} else {
-					completeRequest(req, codes.Internal, fmt.Sprintf("OVMS model unload failed. state: %s", modelStatus.State))
+				// abort any pending load requests for this model
+				if requestMap[req.modelId] != nil {
+					mm.log.V(1).Info("Aborting request due to subsequent unload", "model_id", requestMap[req.modelId].modelId, "request_type", requestMap[req.modelId].requestType, "context_error", requestMap[req.modelId].ctx.Err().Error())
+					completeRequest(requestMap[req.modelId], codes.Aborted, "Aborting due to subsequent unload request")
+					delete(requestMap, req.modelId)
 				}
```

Comment:
Maybe move this outside (before) so that it's shared between load and unload? I guess may need to change the `select` to an `if` for that

Reply:
Isn't it clearer if the reported errors can be different depending on the request type?

What would be the benefit of moving this out?

Reply:
I guess just because its duplicate logic and has the same purpose, i.e. check if there's an existing request and if so abort it. Not a big deal though just a suggestion!